src/networks.py

```diff
@@ -99,6 +106,87 @@ def forward(self, x):
         x = self.layer4["fc"](x)
         return x
 
+    def cache_activations(self):
+        """
+        Creates forward hooks to cache the intermediate activations for each
+        layer.
+        """
+        self.activation = {}
+
+        def hook(model, inpu, output, name):
+            self.activation[name] = F.relu(output.detach())
+
+        self.layer1.fc.register_forward_hook(functools.partial(hook, name=1))
+        self.layer2.fc.register_forward_hook(functools.partial(hook, name=2))
+        self.layer3.fc.register_forward_hook(functools.partial(hook, name=3))
+        self.layer4.fc.register_forward_hook(functools.partial(hook, name=4))
+        return self.activation
+
+    def calc_sparsity(self, device, n=10000, lim=5, do_print=True):
+        """
+        Diagnostic function. Runs the network on n random data points, and
+        returns the proportion of neurons in each layer that are never
+        activated
+        """
+        if self.input_type == "multi":
+            x = (torch.rand((n, self.out)) - 0.5) * 10
+        else:
+            x = (torch.rand((n, 1)) - 0.5) * 10
+        x = x.to(device)
+        _ = self(x)
+        output = []
+        for i in range(1, 4):
+            activated_neurons = (self.activation[i] != 0).sum(axis=0)
+            num_neurons = (activated_neurons > 0).sum().detach().cpu().numpy()
+            total_neurons = len(activated_neurons)
+            if do_print:
+                print("Layer {} has {} activated neurons out of {}".format(
+                    i, num_neurons, total_neurons))
+            output.append(num_neurons / total_neurons)
+        return output
+
+    def calc_arg_deps(self, device, n=100, show_plot=False, fns=None):
```

Comment:
IMO this and cache_activations should be in utils, since they aren't defining networks. Alternatively you could create a new file called network_utils.py and put them there.